Bump CAPI to v1.9.3

EmilienM opened the pull request:
**What this PR does / why we need it**:

Bump to the latest CAPI:
* Updates `go.mod` with the new tag.
* Run `make modules` in root and `orc/`.
* Run `make generate`.
* Add new API violations from what has been generated.
* Add schemes into predicates (see https://github.com/kubernetes-sigs/cluster-api/pull/11239)
* Import deprecated errors that are used in CAPO and make it explicit that they are deprecated and will be replaced by Conditions when time permits.

Issue #2259

EmilienM committed:
Bump to CAPI 1.9.3

EmilienM commented:
/hold
it's still in beta, but at least we have it ready and we can see how it works in our CI.

EmilienM committed:
controllers: adapt for CAPI 1.9 breaking changes

Add schemes into predicates (see
https://github.com/kubernetes-sigs/cluster-api/pull/11239)

EmilienM commented:
The tests fail because https://github.com/kubernetes-sigs/cluster-api/pull/11317 which deprecates `Failure*` fields:
https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/2251/pull-cluster-api-provider-openstack-test/1856337619219124224/build-log.txt

@lentzi90 @mdbooth we should get rid of the `FailureReason` and `FailureMessage`. Do we want a transition step where we carry this interface in CAPO or switch to terminal failures outside of this PR in preparation of a bump to this version of CAPI?

lentzi90 commented:
I think it depends on how much work it would be to switch. If we can do that pretty easily, let's just get it done. Otherwise, let's ignore the deprecation warning and give ourselves more time to work on it in the mean time.

EmilienM committed:
Import deprecated but needed errors constants from CAPI

Context: https://github.com/kubernetes-sigs/cluster-api/issues/10784
We'll maintain them here for the time being, until we have conditions
replacing these errors.

EmilienM commented:
/hold cancel

EmilienM commented:
/approve

lentzi90 reviewed (commented):
Great job!

Do we want to take this in already or wait for the final release?
One thing I am wondering about is also how the FailureMessage will now work. Previously, CAPI would propagate them up to the CAPI objects, but I'm guessing that would no longer happen in v1.9. That would mean that it is harder for users to detect when something is wrong.
If that it the case, should we rather hold off on this until we can transition to conditions instead?

lentzi90 commented:
Yeah I agree it is good to bump early and test it out properly. Shouldn't be a big deal to switch to conditions either.

EmilienM commented:
@mdbooth Matt, what do you think?

EmilienM commented:
TODO: make changes to release notes generator tool in Makefile, the new version in CAPI 1.9 is a bit different. See how VSphere is doing it.
Reference: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/2289

EmilienM commented:
During the CAPO meeting this week, we agreed we could merge this one and iterate over https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/2290 as soon as possible.

MaysaMacedo commented:
/retitle Bump CAPI to v1.9.3

EmilienM commented on controllers/openstackmachine_controller.go:
```diff
@@ -198,11 +198,11 @@ func (r *OpenStackMachineReconciler) SetupWithManager(ctx context.Context, mgr c
 			&clusterv1.Cluster{},
 			handler.EnqueueRequestsFromMapFunc(r.requeueOpenStackMachinesForUnpausedCluster(ctx)),
-			builder.WithPredicates(predicates.ClusterUnpausedAndInfrastructureReady(log)),
+			builder.WithPredicates(predicates.ClusterPausedTransitionsOrInfrastructureReady(mgr.GetScheme(), log)),
```
From CAPI:

```
This predicate is deprecated and will be removed in a future version,
use ClusterPausedTransitionsOrInfrastructureReady instead.
```

https://github.com/kubernetes-sigs/cluster-api/blob/e89268d1c02480c7fb0e8eea7c9a4ee41cf5934c/util/predicates/cluster_predicates.go#L292-L293

lentzi90 reviewed (commented):
/lgtm
